### PR TITLE
Fixed: Specified key was too long; max key length is 767 bytes for ProductPromoCodeEmail entity.(OFBIZ-5426)

### PR DESCRIPTION
**Pull request by pwverma**
Fixed: Specified key was too long; max key length is 767 bytes for ProductPromoCodeEmail entity.

(OFBIZ-5426)

The problem is in the entity model. An email address should not be used in the primary key - mainly because an email address is case-insensitive. A better design would be to use the email address contact mechanism ID in the primary key.

Done Following:
1. Changed Entity Name from ProductPromoCodeEmail to ProductPromoCodeContMech
2. Related Changes for the entity name change
3. Migration service to migrate old data

Thanks, Leon for the report and Adrian Crum, Jacques Le Roux, Ingo Wolfmayr, Deepak Dixit, Pierre Smits and Gil Portenseigne for the discussion and review.

### CODE REVIEW

**Review by dixitdeepak — APPROVED** [review]
Changes looks good to me.

File: applications/product/src/main/java/org/apache/ofbiz/product/migrate/MigrationServices.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
+package org.apache.ofbiz.product.migrate;
```
Comment (jacopoc) [inline applications/product/src/main/java/org/apache/ofbiz/product/migrate/MigrationServices.java:19]: The license header is missing from this class.

File: applications/product/servicedef/services_migration.xml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    <description>Product Component Services</description>
+    <vendor>OFBiz</vendor>
+    <version>1.0</version>
+
+    <service name="migrateProductPromoCodeEmail" engine="java"
+             location="org.apache.ofbiz.product.migrate.MigrationServices" invoke="migrateProductPromoCodeEmail" auth="true">
+        <description>Copy the ProductPromoCodeEmail entity to ProductPromoCodeContactMech</description>
+        <permission-service service-name="productPriceGenericPermission" main-action="CREATE"/>
+    </service>
+</services>
```
Comment (jacopoc) [inline applications/product/servicedef/services_migration.xml]: Please add a new line here.

**Comment by jacopoc** [comment]
I have added a couple of comments for two minor improvements, but apart from them this pull request looks good to me. Thank you.

**Comment by pwverma** [comment]
Thanks, @jacopoc for the review. I have done the changes, thanks!

**Comment by JacquesLeRoux** [comment]
Hi @verma-pawan ,
Should this not be committed now?

**Comment by pwverma** [comment]
> Hi @verma-pawan ,
> Should this not be committed now?

Hey @JacquesLeRoux 
I hope you are doing good. Yes, it's complete and ready to merge. This ticket is assigned to @dixitdeepak  and I think soon he will commit it :)

**Review by dixitdeepak — APPROVED** [review]

**Comment by dixitdeepak** [comment]
> > Hi @verma-pawan ,
> > Should this not be committed now?
> 
> Hey @JacquesLeRoux
> I hope you are doing good. Yes, it's complete and ready to merge. This ticket is assigned to @dixitdeepak and I think soon he will commit it :)

Hi Pawan,

Please feel free to commit :)

**Comment by JacquesLeRoux** [comment]
Thanks guys

**Pull request merged**